yt_dlp/downloader/mhtml.py

```diff
@@ -0,0 +1,175 @@
+
+        return output.getvalue()
+
+    def real_download(self, filename, info_dict):
+        fragment_base_url = info_dict.get('fragment_base_url')
+        fragments = info_dict['fragments'][:1] if self.params.get(
+            'test', False) else info_dict['fragments']
+
+        ctx = {
+            'filename': filename,
+            'total_frags': len(fragments),
+        }
+
+        self._prepare_and_start_frag_download(ctx)
+
+        extra_state = ctx.setdefault('extra_state', {
+            'header_written': False,
+            'mime_boundary': str(uuid.uuid4()).replace('-', ''),
+        })
+
+        frag_boundary = extra_state['mime_boundary'].encode('us-ascii')
+
+        if not extra_state['header_written']:
+            stub = self._gen_stub(fragments, frag_boundary)
+
+            ctx['dest_stream'].write(
+                b'MIME-Version: 1.0\r\n'
+                b'From: <nowhere@yt-dlp.github.io.invalid>\r\n'
+                b'To: <nowhere@yt-dlp.github.io.invalid>\r\n'
+                b'Subject: yt-dlp downloaded slideshow\r\n'
+                b'Content-type: multipart/related; '
+                b'' b'boundary="%b"; '
+                b'' b'type="text/html"\r\n'
+                b'\r\n'
+                b'--%b\r\n'
+                b'Content-Type: text/html; charset=utf-8\r\n'
+                b'Content-Length: %u\r\n'
+                b'\r\n'
+                b'%b'
+                % (frag_boundary, frag_boundary, len(stub),
+                    stub.encode('utf-8'))
+            )
+            extra_state['header_written'] = True
+
+        for i, fragment in enumerate(fragments):
+            if (i + 1) <= ctx['fragment_index']:
+                continue
+
+            fragment_url = urljoin(fragment_base_url, fragment['path'])
+            success, frag_content = self._download_fragment(ctx, fragment_url, info_dict)
+            if not success:
+                continue
+
+            mime_type = b'image/jpeg'
+            # XXX: check extension instead? HTTP headers?
+            if frag_content.startswith(b'\x89PNG\r\n\x1a\n'):
+                mime_type = b'image/png'
+            if frag_content.startswith((b'GIF87a', b'GIF89a')):
+                mime_type = b'image/gif'
+
+            frag_header = io.BytesIO()
+            frag_header.write(b'--%b\r\n'
+                % (frag_boundary,))
+            frag_header.write(b'Content-ID: <%b>\r\n'
+                % (self._gen_cid(i, fragment, frag_boundary).encode('us-ascii'),))
+            frag_header.write(b'Content-type: %b\r\n'
+                % (mime_type,))
+            frag_header.write(b'Content-length: %u\r\n'
+                % (len(frag_content),))
+            frag_header.write(b'Content-location: %b\r\n'
+                % (fragment_url.encode('us-ascii'),))
+            try:
+                frag_header.write(b'X.yt-dlp.Duration: %f s\r\n'
+                    % (fragment['duration']))
+            except KeyError:
+                pass
```

Comment:
I wasn’t entirely sure whether to support streams of duration-less slides. I thought there might be a use case for it, but now that I think of it, it probably doesn’t make much sense.

Though the stub is actually the least important part of this.

Reply:
What I meant is that you are calling `fragment['duration']` everywhere else (instead of `.get`) making it a necessary field by design. So catching `KeyError` here does not make any sense